Power limiting using tMAX with motor resolver speed

ajmel opened the pull request:
First attempt at implementing a limit on power by relating the power limit and resolver speed.

Commit:
First attempt at power limiting testing on stands (cut to zero torque)

Commit:
Adding openOCD directory, adjusting Makefile to point to openOCD to flash on WSL

ajmel committed:
Use openOCD properly, with envvars

ajmel committed:
Updated controls with hinge limiter for power limiting

ajmel committed:
Removed openOCD folder

ajmel committed:
Corrected typos from PR

ajmel committed:
Update controls.c

Very first implementation of power limiter based on motor resolver speed.

ajmel committed:
Update to prevent division by zero

Updated power limiting torque to prevent division by zero

ajmel committed:
Updated tMAX estimate

Updated tMAX by fixing division by zero error, fixed conditional on mc_readings.speed (negative is forward)

ajmel committed:
Added monitoring for tMAX for power limiting

Added CAN message (tMAX) for monitoring power limiting

daniwhite commented on src/vcu/src/controls.c:
```diff
@@ -154,6 +165,21 @@ static int32_t get_regen_torque() {
+static int32_t get_power_limited_torque(int32_t pedal_torque) {
+    if (mc_readings.speed < 0) { //prevent division by zero, make sure we are spinning (negative is forward)
+    
+      int32_t tMAX = power_limit/(abs(mc_readings.speed)*6.28/60); //convert RPM to rad/s with 2pi/60
```
So in general on MY18 you should not be using floats -- do something like this instead:
`int32_t tMAX = power_limit/(abs(mc_readings.speed)*628/6000)`
Also remember that the torque command is in units of deci-newton meters, not Nm. So you should multiply by 10.

daniwhite replied:
(MY19 has hardware floating point support so you don't have to worry about shenanigans like this)

daniwhite commented on src/vcu/src/controls.c:
```diff
@@ -80,7 +87,7 @@ void execute_controls(void) {
     // }
     if (!control_settings.using_voltage_limiting) voltage_limited_torque = torque_command;
-
+  
```
Be careful about little whitespace things like this--it clogs up the diff and also you just generally want to keep it consistent. Make sure empty lines contain no whitespace.
What editor are you using?

ajmel replied:
Currently using Visual Studio Code, unsure why I keep getting whitespace like that because in my editor it looks fine

daniwhite commented on src/vcu/Makefile:
```diff
@@ -122,7 +122,7 @@ C_STD = gnu99
 MOUNT = /Volumes/NODE_F207ZG/
 UART_BAUD = 115200
-COMPORT ?= $(word 1, $(wildcard /dev/tty.usbserial-*) $(wildcard /dev/tty.usbmodem*) $(wildcard /dev/ttyUSB*) $(wildcard /dev/ttyACM*))
+COMPORT ?= /dev/ttyS4 #$(word 1, $(wildcard /dev/tty.usbserial-*) $(wildcard /dev/tty.usbmodem*) $(wildcard /dev/ttyUSB*) $(wildcard /dev/ttyACM*))
```
You should not commit this change. Also this wildcard thing should have worked for you--did you try it and it couldn't find your com port?

ajmel replied:
I was working with Rod on this. Can't remember what happened. Will try to flash with the previous code and let you know if it works or not.

ajmel replied:
I've now tested that this works.

daniwhite commented on can_spec_my18.yml:
```diff
@@ -1536,3 +1536,13 @@ buses:
+        period: 25 ms
+        segments:
+          tMAX:
+            unit: Nm
```
You shoud make this match the units of the torque command (dNm). Also 8 bits is almost certainly too small--only goes up to 2^8=256, and if you fix the units that will not be enough space. You have plenty of space in this message right now--might as well make it an `int32_t`.

daniwhite commented on src/vcu/src/can_handles.c:
```diff
@@ -293,6 +297,14 @@ void handleDashRequest(Frame *msg) {
   control_settings.active_aero_enabled = unpacked_msg.active_aero_enabled;
 }
 
+void handlePowerLimMonitoring(Frame *msg) {
```
You don't need this -- this is just for receiving messages, but you'll only be sending this type of message.

daniwhite commented on src/vcu/src/can_handles.c:
```diff
@@ -340,11 +352,18 @@ void send_VCUControlsMonitoring(void) {
+void send_PowerLimMonitoring(void) {
+  LIMIT(can0_PowerLimMonitoring);
+
+  can0_VCUControlsMonitoring_Write(&power_lim_settings);
```
Change this function to match your message name as well: `can0_PowerLimMonitoring_Write`

daniwhite commented on src/vcu/src/controls.c:
```diff
@@ -100,6 +107,10 @@ void execute_controls(void) {
       min_sensor_torque = temp_limited_torque;
     }
 
+    if (power_limited_torque < min_sensor_torque) {
```
You may want to add a boolean to easily toggle this on and off, but not necessary right now

daniwhite reviewed (changes requested):
This looks pretty good alvaro! I'm glad you were about to get the can message set up. There were just a few things I wanted to point out that will cause trouble if you don't fix them.

ajmel committed:
Fixed tMAX calculation and tMAX CAN message

tMAX calculation is now integer and outputs in dNm, fixed bit length for tMAX message from 8 to 32 bit

ajmel committed:
Fixed whitespace

Fixed whitespace on controls.c

ajmel committed:
Actually fixed whitespace

Fixed whitespace

ajmel committed:
Resolved COMPORT
